### PR TITLE
Cleanup property name computation

### PR DESCRIPTION
**Pull request by tautschnig**
This PR combines multiple cleanup steps (apologies for mixing them, but they depend on each other). Review commit-by-commit.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]
I was contemplating storing the property ID in the goto step, but this works as well. Can we clarify in a comment the difference between property ID and assertion ID? - this seems still vague.

**Review by peterschrammel — CHANGES_REQUESTED** [review]
The property id vs assertion id naming needs to be fixed.

**Comment by tautschnig** [comment]
> The property id vs assertion id naming needs to be fixed.

Any concrete proposals beyond adding a comment? My thinking was that an assertion id generalises a property id, but I'm also ok to use "property" id everywhere.

**Comment by peterschrammel** [comment]
The inconsistency comes from the following:
1. There are assertions that are in the goto program whose property ID is already attached to the source location of the corresponding assert instruction.
2. There are assertions that are not associated with an assert instruction in the goto program.

To accommodate the latter a `comment` field has been added to `SSA_stept` in order to hold the property description (which for the assertions of type 1 is also stored in the source location and just copied over to SSA_step in goto_symex). However, the propery_id is not stored in the SSA_stept, but later generated when printing (!) the goto_trace.

**Comment by peterschrammel** [comment]
What we need is a consistent interface in `goto_trace_stept` and `SSA_stept` to get the property ID (and also its description aka 'comment') associated with an assertion. I think that is what your solution offers, but I don't think that it needs a new name.

Besides, storing the property_id in SSA_stept and goto_trace_stept alongside the `comment` would also make it more consistent and avoid regenerating the property ID in various places.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20b1c1d0683b4303b20d9d4edc13d1a212f610b2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97920470

**Comment by smowton** [comment]
A couple of the callers are clearly expecting a non-nil assertion ID -- could I suggest a `try_get` and `get` pair, where the latter asserts we found an ID?

**Comment by tautschnig** [comment]
@peterschrammel I have tried to implement your suggestions - is that what you were looking for?

@smowton I have clarified the interface contract, in a separate commit.

**Review by smowton — APPROVED** [review]

File: src/goto-symex/symex_target_equation.cpp [hunk 1]
```diff
@@ -1026,7 +1026,7 @@ irep_idt symex_target_equationt::SSA_stept::get_property_id() const
   }
   else
   {
-    // return empty
+    UNREACHABLE;
```
Comment (peterschrammel) [inline src/goto-symex/symex_target_equation.cpp:999]: Just a comment, no action required: If we want to introduce unwinding assertions for recursions then there would be another case here.

**Review by peterschrammel — APPROVED** [review]
Thanks. That looks clearer now.

File: src/goto-symex/symex_target_equation.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-symex/symex_target_equation.cpp:999]: Err, we do. I'll fix the code and will add a test.

Comment (tautschnig) [reply src/goto-symex/symex_target_equation.cpp:999]: (Done)

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 42ea75481616df017bb285783e392e26cdf4fd89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97927228
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 212c793855b6a83d8eea337163f506d5a5d25e01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97937840

**Pull request merged**